src/ORM/DataList.php

```diff
@@ -1128,41 +1136,59 @@ private function fetchEagerLoadHasOne(
             if (is_array($parentData)) {
                 // $parentData represents a record in this DataList
                 $hasOneID = $parentData[$hasOneIDField];
-                $fetchedIDs[] = $hasOneID;
-                $addTo[$hasOneID][] = $parentData['ID'];
+
+                if ($hasOneID > 0) {
```

Comment:
Why was this condition added? It doesn't seem related to polymorphism.

Reply:
No and yes. Entries with 0 should not trigger a query. Especially for polymorph entries this is a problem because the "HasOneClass" field is mostly not empty and can contain any random class.

Reply:
Please either
1. remove these checks (as they aren't related to the problem this PR aims to fix) and open a separate issue and PR for them, OR
2. Change them to `if ($hasOneID)` which reflects `DataObject::getComponent()` and allows for people to change the implementation of primary key to allow non-integer values. Also, if there's _no_ ID, make sure we're adding a `Injector::inst()->create($class)` record (for non-polymorphic relations) or `null` (for polymorphic relations) so that this check doesn't have to happen again in `DataObject::getComponent()`.

Option 1 will make it easier and thus faster to get this bug fix reviewed and merged.

Reply:
Do the same with the other identical check below.

Reply:
It is related. Without this check a empty polymorphic has_one will trigger a query against random tables, because the `*Class` enum will not be empty most of the time ...

Reply:
> Also, if there's no ID, make sure we're adding a Injector::inst()->create($class) record (for non-polymorphic relations) or null (for polymorphic relations) so that this check doesn't have to happen again in DataObject::getComponent().

I could do that, but it would be a change:

```
        // NOTE: Unlike the other relation types, we don't have to explicitly fill empty DataObject records
        // into the has_one components - DataObject does that for us in getComponent() without any extra
        // db calls.
```

`null` for polymorph entries would even be a breaking change: https://github.com/silverstripe/silverstripe-framework/issues/11165

Reply:
> DataObject does that for us in getComponent() without any extra db calls.

I wonder if I wrote that 😂 Okay yeah if that's being handled already then I'm find with it. The only change that's needed then is to not assume the ID is an int.

> This would be nice. But currently not really possible

It is possible, it just requires a lot of work. The more times we assume an int, the more work it will be. We should aim to reduce the number of times we assume the ID is an int rather than increasing it.

> I will change the check against empty value.

Yes please. A simple falsy `if (!$hasOneID)` is sufficient and is used elsewhere.